### PR TITLE
Fixed #28231 -- Doc'd that QuerySet.bulk_create() casts objs to a list.

### PR DESCRIPTION
**Pull request by botondus**
https://code.djangoproject.com/ticket/28231

### CODE REVIEW

**Comment by timgraham** [comment]
It's probably worth mentioning the workaround for high memory consumption as described in https://github.com/django/django/pull/8540#issuecomment-305955006.

**Comment by botondus** [comment]
@timgraham I've added the ``itertools.islice`` example now

**Comment by orf** [comment]
Does that example work as expected?

```python
>>> list(itertools.islice(range(40), 0, None, 10))
[0, 10, 20, 30]
```

Unless I'm completely misunderstanding something, based on the example I would expect it to return `[[0,1,2,3,4,5...], [10,11,12,13,14...], [20,21,22,23....]]`.

**Comment by botondus** [comment]
@orf You're right, the example is wrong. The last parameter for ``islice`` is actually a step parameter. Thanks for catching this, I should've tested it myself.

**Comment by orf** [comment]
There is a nice one line generator expression that I often use, which is on
stack overflow. I will try and find it when I get back to the office. That
could be a simple replacement.

**Comment by botondus** [comment]
There are a number of ways to go about it but unfortunately each is quite a bit more involved than the original (incorrect) example.
This one is a bit verbose but quite easy to read:
```python
from itertools import islice

def split_every(batch_size, iterable):
    i = iter(iterable)
    batch = list(islice(i, batch_size))
    while batch:
        yield batch
        batch = list(islice(i, batch_size))

for batch in split_every(batch_size, iterable):
    Entry.objects.bulk_create(batch, batch_size)
```
Another way that is more compact but perhaps less readable:

```python
from itertools import islice, takewhile, count

batched = takewhile(bool,
                        (list(islice(iterable, batch_size)) for _ in count(0)))
for batch in batched:
    Entry.objects.bulk_create(batch, batch_size)
```

**Comment by orf** [comment]
Couldn't you just do `(queryset[I:I+size] for I in range(0, qs.count(),
size))`?

It requires an upfront count, which is an extra query, but without this you
still need an extra query before you know your at the end of the result set.

**Comment by botondus** [comment]
@orf `bulk_create` works with an iterable of model objects, not necessarily a queryset. In fact I'd imagine it would be uncommon to pass it a queryset since you're trying to insert new data into the database.
In any case here we're looking at the case when a very large iterable is passed to `bulk_create`

**Comment by orf** [comment]
Oh sure, sorry didn't see the context in my email client. Yeah that is
pretty tricky!

**Comment by botondus** [comment]
@timgraham @orf I went ahead and added the workaround which uses `itertools.takewhile`. It seems to me it's a pretty good compromise between verbosity and readability.
Thanks again @orf for catching the incorrect example.

**Comment by timgraham** [comment]
I'm not familiar with the itertools functions in the example but it doesn't look very readable to me. I don't have something better to suggest  right now but I feel like documenting that pattern is inviting a ticket asking why Django doesn't make the use case simpler.

**Comment by timgraham** [comment]
@charettes, do you have an opinion about this?

File: docs/ref/models/querysets.txt [hunk 1]
```diff
@@ -1951,6 +1951,20 @@ This has a number of caveats though:
...
+  case then you can evaluate only a batch at a time by using
+  ``itertools``::
+
+    from itertools import islice, takewhile, count
+
+    iterable = iter(objs)
+    batches = takewhile(bool,
+                        (list(islice(iterable, batch_size)) for _ in count(0)))
+    for batch in batches:
+        Entry.objects.bulk_create(batch, batch_size)
```
Comment (charettes) [inline docs/ref/models/querysets.txt]: What about

```python
iterator = iter(objs)
while True:
    batch = list(itertools.islice(iterator, chunk_size))
    if not batch:
        break
    Entry.objects.bulk_create(batch, batch_size)
```

Comment (charettes) [reply docs/ref/models/querysets.txt]: `iter` is not needed if `objs` is a generator.

Comment (timgraham) [reply docs/ref/models/querysets.txt]: Thanks Simon, I pushed an updated example based on your suggestion.

File: tests/bulk_create/tests.py [hunk 2]
```diff
@@ -121,11 +121,15 @@ def test_batch_same_vals(self):
         self.assertEqual(Restaurant.objects.count(), 2)
 
     def test_large_batch(self):
-        with override_settings(DEBUG=True):
-            connection.queries_log.clear()
-            TwoFields.objects.bulk_create([
-                TwoFields(f1=i, f2=i + 1) for i in range(0, 1001)
-            ])
+        from itertools import islice
```
Comment (charettes) [inline tests/bulk_create/tests.py]: Could be moved to the top.

File: tests/bulk_create/tests.py [hunk 3]
```diff
@@ -121,11 +121,15 @@ def test_batch_same_vals(self):
         self.assertEqual(Restaurant.objects.count(), 2)
 
     def test_large_batch(self):
-        with override_settings(DEBUG=True):
```
Comment (charettes) [inline tests/bulk_create/tests.py]: So weird that we were overriding  `DEBUG=True` here without ever looking up `connection.queries` later on.

File: django/db/models/deletion.py [hunk 4]
```diff
@@ -284,8 +284,8 @@ def delete(self):
 
             # update fields
             for model, instances_for_fieldvalues in self.field_updates.items():
-                query = sql.UpdateQuery(model)
                 for (field, value), instances in instances_for_fieldvalues.items():
+                    query = sql.UpdateQuery(model)
```
Comment (charettes) [inline django/db/models/deletion.py]: That seems to be a `git am` leftover from [#29016](https://code.djangoproject.com/ticket/29016).

File: tests/bulk_create/tests.py (hunk 3 above)
Comment (timgraham) [reply tests/bulk_create/tests.py]: Oops, I didn't mean to commit any test changes.

**Pull request merged**